### ISSUE TITLE
How to convert Open3d camera Pose to unity

### ISSUE THREAD

**Issue by 88ili88** [isl-org/Open3D#4641]
### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](http://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](http://www.open3d.org/docs/release/) and the [latest documentation](http://www.open3d.org/docs/latest/) (for `master` branch).

### My Question

Greetings!
I've been running thru an issue where I do a camera extrinsic calibration in Open3D with global registration and ICP, the point clouds from each camera get aligned correctly but when I pass the camera pose to unity I get a really bad result. The Open3D Transform seems fine i already inverted the Y axis due to difference in the coordinate system but i can't figure out where the variation error is coming from, is it just beacause of impresicion or I'm missing something. 

I invert the transform so i can extract the rotation matrix and the translation vector from the pose. My aproach is to get the rotation and the translation  and assign them to their correspondent kienct object in unity. The main kinect stays in the coordinates origin and then i apply the transformations to each camera referencing the main one. I'll attach some data and captures of the set up. 

![image](https://user-images.githubusercontent.com/95495428/150535084-a67d634c-c684-42e0-8291-cc67d66a0055.png)

right_transformation
array([[ 0.96518757, -0.24386752, -0.09455995, -0.30049289],
       [ 0.14996032,  0.21974007,  0.9639638 , -0.97258158],
       [-0.21430085, -0.94458612,  0.24866084,  0.60323834],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
-----------------------------------------------------------------------
**rotation angles right: [ 75.53550248   5.42599294 -14.17977523] 
translation right: [[0.56515514 0.71024525 0.75911709]]** 

left_transformation
array([[ 0.9690198 ,  0.23938183, -0.06080263, -0.36967974],
       [-0.15101459,  0.37945827, -0.91280119,  1.0112917 ],
       [-0.19543595,  0.89370452,  0.40385273,  0.61812839],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])   
--------------------------------------------------------------------
**rotation angles left: [-66.13385402   3.48588443  13.87624411] 
translation left: [[ 0.63175131 -0.84767252  0.65099793]]**

The translation in unity kinda makes sense with the actual set up, but can't get a correct result.

![setup](https://user-images.githubusercontent.com/95495428/150542619-e63343d5-02a2-4cd4-a278-8ea99bcb2f92.jpg)

When I manually calibrate the cameras in unity i get this vectors, and I really can't figure out if there's any correlation with the ones that i mentioned previously.

**rotation angles right: [ 72.861, 7.016,  20.971] 
translation right: [0.3640004, 0.9750017, 0.832]** 

**rotation angles left: [-76.012, 22.273, -35.441] 
translation left: [ 0.331, -1.025,  0.556]**  

![image](https://user-images.githubusercontent.com/95495428/150541953-d363f2b1-7c86-40eb-8d78-cccabdbe2946.png)

Any help or tip about what's happening would be greatly appreciated since i've been stuck with this for a long time, 
Thanks in advanced!

**Comment by 88ili88** [comment]
issue #2218  has a similar problem but it's from unity to open3d.

**Comment by theNded** [comment]
Please refer to #2218 and invert the coordinate transform. As clearly noted in that issue, the operation merely reverts y-axis, either unity to o3d or the way around.

**Comment by theNded** [comment]
Ah I see you have inverted axses. How did you do that, could you please put more details?

**Comment by 88ili88** [comment]
Hey, first of all, thanks for the fast reply!
For the axis inversion, I just multiplied by -1 the Y-axis of the translation vectors that I get from o3d.
In regard to the figures, 2 is what I get when I directly input the vectors that I get from o3d to unity, it looks completely off.
After manually calibrating the cameras in unity I get 3 and 4, which gives me the second set of vectors that should be the right answer but I can't find any correlation between the first set of vectors (o3d) and the second set (unity).

**Comment by theNded** [comment]
> For the axis inversion, I just multiplied by -1 the Y-axis of the translation vectors that I get from o3d.

This is incorrect from the beginning. The change of the coordinate system also affects the rotation. #2218 has given detailed transformation conversion. Please apply changes to the rotation as well.

**Comment by 88ili88** [comment]
![image](https://user-images.githubusercontent.com/95495428/150961973-85fc889f-6b8f-4391-86b2-233d2b60125a.png) 

After computing the inverse transform of the point cloud to get the camera pose I proceed as shown in #2218 multiplying by -1 the whole second row of the pose (I don't know why he is multiplying the third one as this corresponds to Z and not Y).
 ![image](https://user-images.githubusercontent.com/95495428/150961912-ec7ffc74-3887-48f8-8e7f-1d4060d0810e.png) 

And finally I get these vectors that should be similar to the second set of vectors, that I got from unity, but that is not the case.
 ![image](https://user-images.githubusercontent.com/95495428/150962203-b31a1c40-46e1-480e-bfa7-2f18fc386242.png)
This gives me a even worse result. 
![image](https://user-images.githubusercontent.com/95495428/150963743-b72dd2fc-065b-41f4-b67a-891cb6b87cad.png)
 I attached some code captures in case u notice something that I'm missing.

**Comment by theNded** [comment]
No, the trick he did was `mat = glm.rotate(mat, angle, [x, -1.0 * y, z])`. `glm` is a fast math lib for graphics and there could be other equivalents for the equation. The coord transfer you referred to could be some system specific transform of his own code.

**Comment by 88ili88** [comment]
But as I showed you, im already inverting the second row of my pose isn't that what's im supposed to do? 
![image](https://user-images.githubusercontent.com/95495428/150975122-ac2e6daa-b436-49c9-8f38-d2d17fe6ceed.png)

**Comment by theNded** [comment]
AFAIK this is incorrect. You need to invert the column as well because the basis also changes.
Technically it can be done by `r.inv() @ R @ r` where `r = [[1,0,0],[0,-1,0],[0,0,1]]`, but since you are not sure of the coordinate systems and have quaternions, Euler angles, transformations, and inversions back and forth, I would suggest you strictly go through #2218 and/or use glm as suggested.

**Comment by theNded** [comment]
Also I'm not sure it is correct to invert the transform. Transform is interpreted as `transformation from camera to world`, which is already its translation and rotation.

**Comment by 88ili88** [comment]
> Transform is interpreted as `transformation from camera to world`, which is already its translation and rotation.
The right_transformation and left_transformation that I mention at the begining of the issue are the pointcloud transformations that align them together, obtained from point2plane ICP, to the best of my knowldge if I inverse this transformation I get the extrensic matrix (rotation and translation) of the cameras

**Comment by theNded** [comment]
Extrinsic matrix is `world to camera` when you apply it to a point. There are different conventions in describing `rotation and translation of the cameras`, but if you need a coordinate system transform, you'd better stick to `camera to world`.

**Comment by 88ili88** [comment]
Hi! First of all, thank you very much for the transformation of coordinate systems write up. I have tried to apply this transformation to the point cloud transform and from there, computing the camera pose, but the result wasn’t the expected.

Now, in order to see exactly what I obtain, I’ve tried to apply the coordinate system transformation and visualize the coordinates using o3d.geometry.TriangleMesh.create_coordinate_frame(). 

To do so, I have defined an identity matrix and applied the transformation of the system coordinates as you mentioned (I attach the code below). After visualizing the coordinates (before and after the transformation) this was the result:
![image](https://user-images.githubusercontent.com/95495428/151991019-14731a37-9c6c-49c7-8917-a75d8d7c3f23.png)
(The two coordinates are on top of each other)

To dissipate any doubts I have applied a translation in the y axis. We can verify that the y axis is inverted but there’s no rotation applied.
![image](https://user-images.githubusercontent.com/95495428/151991104-4b57ed1b-3a6e-4791-a5ba-854ab85d28e1.png)
Code:
![image](https://user-images.githubusercontent.com/95495428/151991183-7056a259-3bbf-4814-a653-6f115811aa56.png)
Then, instead of applying the transformation of the system coordinate to the given transformation matrix, I have inverted the y axis of the translation and I have applied a second rotation to the coordinate mesh, corresponding to the rotation of the coordinate system. And the result was the expected!

I have attached the code and the result below:
![image](https://user-images.githubusercontent.com/95495428/151991252-372a95b0-336c-4cab-b9e1-a7cd9bc8b285.png)

After trying a few things, I have seen that I could obtain the same result by making the following operation:   (I don’t understand why, as the math you showed made very sense).

So I tried to do so and see the results in Unity, but the reconstruction wasn’t correct. Also, I applied first the transformation of the system coordinates and then, I extracted the rotation and translation of the camera, as you suggested too. I attach the code.
![image](https://user-images.githubusercontent.com/95495428/151991541-c8c157c3-690b-4553-a205-56785b86db77.png)
![image](https://user-images.githubusercontent.com/95495428/151993021-57d87ab4-b6f5-4a24-b8a6-7d3d5d28afa5.png)
![image](https://user-images.githubusercontent.com/95495428/151993062-7ac15ed9-48ac-4215-939b-32ce5d676f15.png)
![image](https://user-images.githubusercontent.com/95495428/151993129-4b2ca1de-18ec-4da8-959b-1fffbdff3703.png)


This time, to extract the rotation and translation of the camera from the transformation of the point cloud I didn’t use the inverse transform, but the results are the same. If you want me to add more detail on the maths, tell me so!

I don’t understand which could be the problem. Any suggestion will be appreciated! 
Thanks!

**Comment by Zulex** [comment]
Don't have the solution, but as I am trying to achieve something similar hereby some input:

Have you checked row-major, column major differences? As far as I understand, we have to negate the second column and not the second row, as that is the one storing the y axis information. I could also not find any information of the coordination frame open3D is using next to the other issue mentioned here. So if you have a link stating the open3D coordinate frame, please let me know.

Next to that, it seems that issue 2218 seems to negate the second column and row. If I am correct the glm function he calls is negating the second column and then afterwards he is also negating the second row manually. (which leaves [2,2]  exactly the same as before. However he returns that value twice, so not sure if he negates both or already uses the matrix with only the column negated.

**Comment by BrightLight1115A** [comment]
> 

Hi there.
Nice to meet you.
I am Unity developer and interested in use of Open3D library in Unity.
You seems already using Open3D library in Unity.
So I would appreciate if you let me know how to import Open3D library.
Currently, I imported some Open3D project but I can't sure that it is right project for Unity.
I look forward to hearing from you.

Thank you.
